server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -313,7 +314,8 @@ public IndexShard(
                         indexSettings,
                         UNASSIGNED_SEQ_NO,
                         globalCheckpointListeners::globalCheckpointUpdated,
-                        threadPool::absoluteTimeInMillis);
+                        threadPool::absoluteTimeInMillis,
+                        retentionLeaseSyncer);
```

Comment:
I think this callback needs to acquire a permit before being invoked.

Reply:
I opened #37440 so that this will be simple (and I have another use-case in mind).